Remove building wp_template_parts variation call for frontend

kt-12 opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: [https://core.trac.wordpress.org/ticket/59969](https://core.trac.wordpress.org/ticket/59969)

Gutenberg PR - https://github.com/WordPress/gutenberg/pull/56952

For TT4, running 500 rounds twice and taking the mean.

| Title                                      | Mean (Trunk) | Mean (PR) |
|--------------------------------------------|--------------|-----------|
| Response Time (median)                     | 403.305      | 398.08    |
| wp-load-alloptions-query (median)          | 3.19         | 3.225     |
| wp-before-template (median)                | 171.795      | 154.855   |
| wp-before-template-db-queries (median)     | 0            | 0         |


---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

kt-12 committed:
remove varation

joemcgill commented on src/wp-includes/blocks/template-part.php:
```diff
@@ -272,12 +272,12 @@ function build_template_part_block_variations() {
+		'render_callback' => 'render_block_core_template_part',
 	);
+	if ( current_user_can( 'edit_theme_options' ) ) {
+			$args['variations'] = build_template_part_block_variations();
```
I wonder if the capability check should be moved into the `build_template_part_block_variations` function itself and return an empty array, or `null` if false. I also wonder if this is even the appropriate capability check, since there might be valid use cases where someone needs to get a list of the registered variations even though they can't edit theme options directly (though, I don't have an idea what that use case might be at the moment).

kt-12 replied:
Yes @joemcgill , we can move it to `build_template_part_block_variations`. I was thinking what if we use `is_user_logged_in` instead. This way we will only restrict it to all the users that are logged in without touching a specific capability.

joemcgill replied:
I don't mind the idea of using the capabilities API as long as we're sure we're identifying the correct cap needed to make use of these variations. Otherwise, even logged in users without any editing capabilities will get these variations on the front-end. 

Reading back through [the original thread](https://github.com/WordPress/gutenberg/issues/45601) about this in the Gutenberg repo, it seems like this is only used when editing templates, but @Mamaduka preferred a more declarative approach at the block API level, rather than needing to handle this on a block-by-block basis. That seems to be what @spacedmonkey was attempting to solve in https://github.com/WordPress/wordpress-develop/pull/4857. I tend to agree, and think we have time to be thoughtful about how to declaratively load variants only when needed, or make the variant property a callback that is only loaded when the variants are needed. What do you think?

Mamaduka replied:
Thanks for the ping, @joemcgill!

I still think a declarative approach for lazy-loading would be best suited here. There are other blocks that generate variations on the server where the capabilities approach might not be applicable.

@spacedmonkey, I know you're taking a well-deserved break from contributing, but if you could share your notes for https://github.com/WordPress/wordpress-develop/pull/4857, maybe someone could pick it up again.

spacedmonkey replied:
I don't like this change. It feels like a big backwards compablity break. In https://github.com/WordPress/wordpress-develop/pull/4857, I worked on using a magic getting on class property to lazily load block varients. This isn't great, as it has it's own problems, but it better than not loading them at all unless logged in, which will have unknown side effects.

spacedmonkey replied:
Side note, this change should be done in gutenberg, as this files are versioned there.

joemcgill replied:
@spacedmonkey and @Mamaduka, re:

> Side note, this change should be done in gutenberg, as this files are versioned there.

This PR, and the related Trac ticket were specifically created based on the conversation [the two of you had here](https://github.com/WordPress/gutenberg/issues/45601#issuecomment-1552441434). In which, @Mamaduka said that we needed these changes to be made in WP. Could one of you two clarify which changes you were referring to?

Mamaduka replied:
@joemcgill, that’s for changes required to lazy-load variations property. We can’t override that from the plugin.

kt-12 replied:
I have modified the code so that variations are loaded when required (the original approach @spacedmonkey  had implemented).

@Mamaduka one of the issues I am facing is how to visually test the functionality. Right now even if I remove the variation code entirely I see variation on the editor.

By analysing the code so far I know, 

- that variation inside `get_block_editor_server_block_settings`  is not needed in frontend and
- get_block_editor_server_block_settings is used to preload the block setting via inline script for ->
[site-editor](https://github.com/WordPress/wordpress-develop/blob/0ce733c78f86c4ea06dd0dde329fcca571179f50/src/wp-admin/site-editor.php#L131), [widgets-form-blocks](https://github.com/WordPress/wordpress-develop/blob/0ce733c78f86c4ea06dd0dde329fcca571179f50/src/wp-admin/widgets-form-blocks.php#L51), [edit-form](https://github.com/WordPress/wordpress-develop/blob/0ce733c78f86c4ea06dd0dde329fcca571179f50/src/wp-admin/edit-form-blocks.php#L110), [wp-customize-widgets](https://github.com/WordPress/wordpress-develop/blob/0ce733c78f86c4ea06dd0dde329fcca571179f50/src/wp-includes/class-wp-customize-widgets.php#L866)

Visually, I was able to confirm that these inline values were the same after our change, so nothing should be breaking. But, I am not sure how and when the value is getting consumed and how to test the functionality.

@joemcgill Another issue is that even though this approach only loads variation in the editor it is not truly lazy loading. It's still loaded in a blocking way on the editor side. However, @spacedmonkey  has pointed out a way around 3 years back on a different thread to make this entirely lazy loading but it was never explored - https://github.com/WordPress/gutenberg/issues/22812

joemcgill reviewed (commented):
Left a few questions, but this would be a simple updated if it doesn't cause any unforeseen side effects.

kt-12 committed:
introduce get all variation

kt-12 committed:
cs fixes

kt-12 committed:
cs fixes

joemcgill commented:
Thanks for the latest update @kt-12. Going to break out of the inline thread and respond to [your last update](https://github.com/WordPress/wordpress-develop/pull/5718#discussion_r1411676946) here.

This approach is a bit better, because it allows variations to only be loaded in `get_block_editor_server_block_settings()`, but it still seems like a workaround and not a true API, since this will cause the registered block properties to be different based on whether that filter happens to be called, which also makes registered blocks mutable and thus non-cacheable (or at least more challenging). I do think that ensuring variations show up in the API responses would be a requirement, even if we did temporarily move variant registration for this block to only `get_block_editor_server_block_settings()` and the API, though.

What would you think about updating the block registration schema so that you could pass a callable as the `variations` property, which would be called only when variations are needed? @Mamaduka suggested something similar in the original Gutenberg thread.

Additionally, I think for consistency, we should apply whatever changes we're making for template part variations to other core blocks that register variations dynamically on the server, e.g. [Navigation Link](https://github.com/WordPress/wordpress-develop/blob/6.4.1/src/wp-includes/blocks/navigation-link.php#L363-L369), and [Post Terms](https://github.com/WordPress/wordpress-develop/blob/6.4.1/src/wp-includes/blocks/post-terms.php#L106-L112).

BTW, this change does seem to have a measurable impact on server response time on the front end, in my local benchmarking, trunk is loading in `290.79 ms (p50)` while this PR is `242.19 ms (p50)`, which is a sizable improvement.

kt-12 committed:
filter to callback

mukeshpanchal27 commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -113,7 +113,7 @@ class WP_Block_Type {
 	 *
 	 * @since 5.8.0
-	 * @var array[]
+	 * @var array[] | callable
```
```suggestion
	 * @var callable|array[]
```

Nit-pick

joemcgill replied:
We'll also need a `@since` tag that documents which version adds support for `callable` values.

joemcgill reviewed (commented):
This looks pretty straightforward, @kt-12. I assume we also need to do something similar in the other places where WordPress expects to return a block definition that includes the full set of registered variations, like in `WP_REST_Block_Types_Controller` or additionally implement as a getter method in `WP_Block_Type` as @Mamaduka was suggesting [here](https://github.com/WordPress/gutenberg/issues/45601#issuecomment-1503585345).

mukeshpanchal27 commented:
@kt-12 Some unit tests going fails, The `wp-includes/blocks/template-part.php` needs to update in GB first then the changes needs to backport.

kt-12 committed:
Add Variation to block-types rest api

kt-12 committed:
navigation link and post term to block varation to callback

kt-12 commented:
@joemcgill Implemented the same for Navigation-Link and Post-Term and then did some benchmarking.

The following command was run thrice for both Trunk and PR and then Mean (Trunk) and Mean (PR) was obtained 
> npm run research -- benchmark-server-timing --url http://localhost:8889/ -n 500 -c 2

### Performance Improvement for TT4 (Tested locally on default homepage)
| Metric                            | Mean (Trunk) | Mean (PR) | Diff        |
|-----------------------------------|--------------|-----------|-------------|
| Response Time (median)            | 397.66 ms    | 362.16 ms | -35.5 ms    |
| wp-load-alloptions-query (median) | 3.23 ms      | 2.96 ms   | -0.27 ms    |
| wp-before-template (median)       | 173.53 ms    | 143.51 ms | -30.02 ms   |

@Mamaduka This PR is purely based on our findings-based interpretation that the dynamically registered variations were only used inside `get_block_editor_server_block_settings`  and `block-types` rest API. We want someone core to Gutenberg to check and confirm if these changes aren't breaking anywhere else.

Furthermore, I have some improvements found on the editor side that could lead to similar performance upgrades on the editor side. However, I don't have the expertise to confirm my findings.

### Remove inline block schema and replace it with prefetch
We have used the following code at [site-editor](https://github.com/WordPress/wordpress-develop/blob/0ce733c78f86c4ea06dd0dde329fcca571179f50/src/wp-admin/site-editor.php#L131), [widgets-form-blocks](https://github.com/WordPress/wordpress-develop/blob/0ce733c78f86c4ea06dd0dde329fcca571179f50/src/wp-admin/widgets-form-blocks.php#L51), [edit-form](https://github.com/WordPress/wordpress-develop/blob/0ce733c78f86c4ea06dd0dde329fcca571179f50/src/wp-admin/edit-form-blocks.php#L110), [wp-customize-widgets](https://github.com/WordPress/wordpress-develop/blob/0ce733c78f86c4ea06dd0dde329fcca571179f50/src/wp-includes/class-wp-customize-widgets.php#L866)
It's a 5-year-old code mostly introduced to handle some edge cases related to third-party blocks and dynamic variations.
```
wp_add_inline_script(
	'wp-blocks',
	'wp.blocks.unstable__bootstrapServerSideBlockDefinitions(' . wp_json_encode( get_block_editor_server_block_settings() ) . ');'
);
```
The above code preloads the schema for all block types. In the default setup, I found it to add 92 block types with it's variations.
I could confirm this by running `wp.blocks.getBlockTypes()` in the browser console for the editor. However, if I comment the above code (I tried on edit-form), reload the editor and then run `wp.blocks.getBlockTypes()` it still loads all the block types, except that some blocks have dynamic information missing in this case. This means that information in inlining is being used but most of the data is just a repeat of what is already available at the editor.

Also, the information in the inline script is only used when a person searches for a block or some similar option and is not needed when the page loads (just my understanding based on what I discovered - this could defer). So we don't need all the information in the first load.

#### Proposal
1. **Replace unstable__bootstrapServerSideBlockDefinitions with a call to block type REST API** [#22812](https://github.com/WordPress/gutenberg/issues/22812)
 The information in `/wp/v2/block-types` is similar to what we are inlining right now. However, it seems that we are not making use of this API endpoint anywhere in the core (not 100% sure). Instead of inlining, we can shift the load to preloading this way we free up the original doc from expensive computation

2. **Only load what is missing instead of looping through all 92 blocks.**
 Most of the information is getting loaded in the editor even without inlining. So we could avoid looping through all the registered block types and inline it entirely. Instead, only inline those blocks that had some dynamic information. Note: This method would yield lesser improvement than the first.

kt-12 commented:
> @kt-12 Some unit tests going fails, The `wp-includes/blocks/template-part.php` needs to update in GB first then the changes needs to backport.

Thanks @mukeshpanchal27. Once I have confirmation on the code here I'll raise another PR for GB. At the moment I am confused as to which will part will go in GB and which will stay here, so focusing more on the problem to solve.

kt-12 committed:
Update src/wp-includes/class-wp-block-type.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

kt-12 committed:
Merge branch 'enhancement/remove-variation-fontend' of https://github.com/10up/wordpress-develop into enhancement/remove-variation-fontend

kt-12 committed:
doc block

kt-12 committed:
dock block

kt-12 committed:
phpcs fixes

kt-12 committed:
Merge branch 'trunk' of https://github.com/WordPress/wordpress-develop into enhancement/remove-variation-fontend

mukeshpanchal27 commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -113,7 +113,8 @@ class WP_Block_Type {
 	 *
 	 * @since 5.8.0
-	 * @var array[]
+	 * @since 6.5.0 Support callback.
```
```suggestion
	 * @since 6.5.0 Added callback support.
```

joemcgill commented on src/wp-includes/blocks/post-terms.php:
```diff
@@ -102,12 +104,18 @@ function register_block_core_post_terms() {
 		array(
 			'render_callback' => 'render_block_core_post_terms',
-			'variations'      => array_merge( $built_ins, $custom_variations ),
+			'variations'      => 'get_variations_core_post_terms',
```
It would be nice to use a consistent naming scheme for all of the variation callbacks in core blocks. Right now you have:

* get_variations_core_navigation_link
* get_variations_core_navigation_link
* build_template_part_block_variations

Personally, I'd go with something like `get_block_core_{block_name}_variations` which is similar to the consistent render callback `render_block_core_{block_name}`.

kt-12 replied:
I was avoiding `build_{block name}` as the navigation link already had a similar function with a different purpose ->`build_variation_for_navigation_link`.

I like the `get_block_core_{block_name}_variations` naming throughout, but we may have to deprecate the function `build_template_part_block_variations` which is old and can be in use in ways we don't know.
For now, I have made everything `build_{block}_block_variations` for consistency.

kt-12 replied:
@joemcgill  I have created a draft PR in GB repo - https://github.com/WordPress/gutenberg/pull/56952/files

There are some CS issues which I think someone from GB would be the best person to address.

joemcgill reviewed (commented):
Thanks @kt-12. This is looking promising. I'd love to get feedback from someone like @Mamaduka or @ockham on the general approach.

One important thing to note is that the changes to individual blocks will need to be made in the Gutenberg repo after support for registering variations as a callable is added. I think it makes sense to show the whole change in this PR in the meantime, but we'd likely only commit the changes to `get_block_editor_server_block_settings()` and the API.

kt-12 committed:
Update src/wp-includes/class-wp-block-type.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

kt-12 committed:
Making name consistant

kt-12 committed:
Merge branch 'enhancement/remove-variation-fontend' of https://github.com/10up/wordpress-develop into enhancement/remove-variation-fontend

spacedmonkey commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -113,7 +113,8 @@ class WP_Block_Type {
 	 * @since 5.8.0
-	 * @var array[]
+	 * @since 6.5.0 Added callback support.
+	 * @var callable|array[]
```
This really feels like a breaking change. This is a public property on a class. I don't think we should do this.

joemcgill replied:
We could consider making this property private and use a `__get()` method to check and see if the block type was registered with a callback stored in a new `$variations_callback` property. If a callback was registered, the magic getter would return the result of the callable. If no callable is registered then it would return the property itself. We would need a `__set()` method to maintain the previous behavior of directly setting the variations property. However, this would mean that if a callback were available, it will be prioritized over any directly set value unless we somehow merged the callback value with any currently set values of the variations prop itself, which seems like it would get. messy.

Mamaduka replied:
I agree with @spacedmonkey; this feels like a breaking change. Calling `$block_type_instance->variations` should return an array.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php:
```diff
@@ -314,6 +313,14 @@ public function prepare_item_for_response( $item, $request ) {
 			}
 		}
 
+		if ( rest_is_field_included( 'variations', $fields ) ) {
+			$field = $block_type->variations;
+			if ( is_callable( $block_type->variations ) ) {
+				$field = call_user_func( $block_type->variations );
+			}
+			$data['variations'] = rest_sanitize_value_from_schema( $field, $schema['properties']['variations'] );
+		}
```
I like this 👍

spacedmonkey reviewed (changes requested):
Changing the type of this property on the block type class feels unwise.

Mamaduka commented on src/wp-includes/blocks/template-part.php:
```diff
@@ -276,7 +276,7 @@ function register_block_core_template_part() {
 		array(
 			'render_callback' => 'render_block_core_template_part',
-			'variations'      => build_template_part_block_variations(),
+			'variations'      => 'build_template_part_block_variations',
```
I think it's okay to have these changes here for now and remove them before merging into the core. Then, we can update the code in the `block-library` package in a backward compatible way. The Gutenberg plugin supports multiple WP versions.

spacedmonkey commented:
I have created my own PR https://github.com/WordPress/wordpress-develop/pull/5825 based on this one. It is very similar but has some improves.

gziolo commented:
> The above code preloads the schema for all block types. In the default setup, I found it to add 92 block types with it's variations.
I could confirm this by running wp.blocks.getBlockTypes() in the browser console for the editor. However, if I comment the above code (I tried on edit-form), reload the editor and then run wp.blocks.getBlockTypes() it still loads all the block types, except that some blocks have dynamic information missing in this case. This means that information in inlining is being used but most of the data is just a repeat of what is already available at the editor.

We could also remove the duplicated information on the client for core blocks, given that the server is the source of truth and it always takes precedence. We need to keep in mind that everything is filterable in WordPress core so every block type can get modified by the site before exposing the final shape to the client. That's why I would prefer to replace `wp.blocks.unstable__bootstrapServerSideBlockDefinitions` with the REST API endpoint call for block type and if necessary preload the response on the server. In reality, the won't be much of a difference for the user as the simply need all the information from the server to have the correct state in the block editor.

That said, the subset of changes for lazy loading block variations that are also discussed in https://github.com/WordPress/wordpress-develop/pull/5825 seems like a nice optimization. If we were to add a new setting like `variations_callback` then we will have to polyfill it in the Gutenberg plugin so it works with all supported versions of WordPress. There is an existing filter that should make it possible:

https://github.com/WordPress/wordpress-develop/blob/e43275b61c6e3a8d5657b6d21eb99bf1bafd52b2/src/wp-includes/class-wp-block-type.php#L524

For older versions of WordPress `variations_callback` would have to be called and stored under `variations` which isn't ideal, but at least for WP 6.5 we will see all the improvements.

kt-12 committed:
callback support with backward compatibility

kt-12 committed:
Allow setting of value. Doc block for variation callback

kt-12 commented:
Update this PR for B/C mainly based on @spacedmonkey's new PR update. 

Below are some changes that were made to the original PR and the reasoning behind it ->

1. In this PR, `$variations` is set to `private` instead of `protected` as both yield different results (added a short snippet below to explain this). At the moment (trunk) final built variations are available to the child class, so we must respect that for B/C, hence `private`. 
2. In the current trunk code, variations are only built once during block registration. The same can be assumed for the updates that we make. So, I have updated the original `$variations` variable once the callback is made, this will prevent expensive callback from being called again and again if accessed multiple times in the same request. However, I have added a filter `allow_variation_callback_recall` which when set to true should allow recalling callback.
3. I believe that even though `$variations` is set as private by nature it is not truly private, so we could avoid issuing a warning for anyone who tries to access them directly instead of via the `get_variations()` method.   

```php
<?php

class a {
    private $a  =  'b'; 
    
    public function __get( $name ) {
		if ( 'a' === $name ) {
			return 'c';
		}
    }
}

class c extends a {
    public function test() {
        print_r( $this->a );
    }
}

$f = new c();
$f->test();  // this gives output as 'c', however, if we use `protected $a` you get the result as 'b'
```
cc: @joemcgill @spacedmonkey

joemcgill commented on src/wp-includes/blocks/navigation-link.php:
```diff
@@ -323,12 +323,11 @@ function build_variation_for_navigation_link( $entity, $kind ) {
+ * @return array
  */
-function register_block_core_navigation_link() {
+function build_navigation_link_block_variations() {
```
Setting aside the fact that these will need to be changed in the Gutenberg repo, rather than directly in core, all of these callback functions should probably be updated to include "core", like `build_core_navigation_link_block_variations`, etc., to avoid potential conflicts with any third party blocks.

gziolo replied:
There is a linting rule in the Gutenberg plugin that should enforce using a proper prefix for the function.

kt-12 replied:
@joemcgill my only concern about changing name is consistency in naming convention. 

Giving that `build_template_part_block_variations` is an older public function, moving to `build_core_*` means we might have to depreciate `build_template_part_block_variations` with a wrapper code calling `build_core_template_part_block_variations`.

joemcgill replied:
Ah, that makes sense. I hadn't thought about this already being in core with this name. It's somewhat unfortunate naming, but probably fine. We can raise this concern in the related Gutenberg PR where these changes will need to be made. Not a blocker for this effort.

joemcgill commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -372,6 +390,11 @@ public function __isset( $name ) {
 	 * @param mixed  $value Property value.
 	 */
 	public function __set( $name, $value ) {
+		if ( 'variations' === $name ) {
```
👍🏻

joemcgill commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -540,4 +563,21 @@ public function get_attributes() {
+	 * @return array[]
+	 */
+	public function get_variations() {
+		if ( isset( $this->variation_callback ) && is_callable( $this->variation_callback )
+			&& ( empty( $this->variations ) || apply_filters( 'allow_variation_callback_recall', false ) )
+		) {
+				$this->variations = call_user_func( $this->variation_callback );
+		}
+
+		return $this->variations;
```
Let's separate out the conditions to make it a bit more readable and to allow us to document any filters we add to this method. 

I don't think that we want to support re-running a variation callback via a filter, as that could cause manually set variations to be overwritten. Instead, I think we could filter the return value so that extenders that want to modify the registered variations can do so after the fact. In doing so, we could pass the original object so that if someone wanted to manually rerun the callback they could. Here's my suggestion:

```suggestion
		if ( empty( $this->variations && is_callable( $this->variation_callback ) ) ) {
				$this->variations = call_user_func( $this->variation_callback );
		}

		/**
		 * Filters the registered variations for a block type.
		 *
		 * @since 6.5.0
		 *
		 * @param array                    $variations   Array of registered variations for a block type.
		 * @param WP_Block_Type $block_type The full block type object.
		 */
		return apply_filters( 'get_block_type_variations', $this->variations, $this );
```

spacedmonkey replied:
What if the callback returns any empty array? That would mean it would run every time?

kt-12 replied:
@spacedmonkey A valid point. It doesn't make sense to run it every time if the callback returns an empty array. To avoid it I have now initialised `$variations` to null; setting it to an empty array only when it is referenced. Now, call to callback also happens only when `$variations` is not set.

@joemcgill  re-running a variation callback via a filter is risky, there is a risk of some plugin accidentally setting it to true, so updated the code the way you suggested.

joemcgill reviewed (changes requested):
Thanks @kt-12. I think this is looking pretty good and your explanation about marking the property as `private` instead of `protected` makes sense to me. I've got some suggestions for how to simplify the getter with an alternate way of allowing extenders to modify variations after the initial callback has been run.

It looks like the PHPUnit tests still need to be updated and we'll want to remove the changes to the blocks directory before committing and open a new PR in Gutenberg to handle those changes there.

spacedmonkey commented on src/wp-includes/blocks/template-part.php:
```diff
@@ -275,8 +275,8 @@ function register_block_core_template_part() {
-			'render_callback' => 'render_block_core_template_part',
-			'variations'      => build_template_part_block_variations(),
+			'render_callback'    => 'render_block_core_template_part',
+			'variation_callback' => 'build_template_part_block_variations',
```
We are going to need some conditional logic here, as these PHP is stored in Gutenberg that supports two versions of WP back.

joemcgill replied:
I'm not sure we need conditional logic here, because this change will only be made in versions of GB or Core that support the `variation_callback` registration arg. What's more important is ensuring that previous versions of GB which build these variations during registration still work as expected once the changes to `WP_Block_Type` are merged in core. I believe our tests will already account for this, since when this is merged we'll still have the previous block definitions in place.

kt-12 committed:
variations set to null initially

joemcgill reviewed (commented):
I like these changes 👍🏻

Can you update the tests to pass and then remove the changes to the actual block files that will need to be handled in Gutenberg? Then I think this is close to being ready for approval.

gziolo commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -540,4 +563,30 @@ public function get_attributes() {
+		 * @param array         $variations Array of registered variations for a block type.
+		 * @param WP_Block_Type $block_type The full block type object.
+		 */
+		return apply_filters( 'get_block_type_variations', $this->variations, $this );
```
I  checked the latest changes in the PR and everything looks great. It’s quite interesting how you managed to go with fully backward compatible approach that works out of the box with things like REST API endpoint.

My only question is related to this new filter as I couldn’t find information what role it plays. What are the use cases it would cover?

kt-12 replied:
@gziolo The main reason to have a filter here is to provide the ability to re-run the callback in case there were any possible changes to the content after the first build of variation.

kt-12 replied:
For example, in  - https://github.com/WordPress/gutenberg/pull/54801, the variation is available after the registration. 

Even though what we are trying to solve is not similar to the issue listed in PR#54801, it did make us think about what if there was a change in the variation that was not present during the call to callback.

gziolo replied:
Interesting, I now can see how that might become useful.

Coincidentally, today I commented on the proposal for building a formal API for registering block variations with PHP: https://github.com/WordPress/gutenberg/issues/47170. Do you have any thoughts on how it all could work together with the changes proposed in this PR. It seems like using the new filter would cover some of the use cases.

kt-12 replied:
Gave a quick look into https://github.com/WordPress/gutenberg/issues/47170. I believe this filter will solve the problem.  The developer can add a function to run the query loop and modify the variations returned by the API. With this hook, the person could append to the existing variations or modify them completely. 

There is yet another approach if the developer wants to replace variations returned by the core completely (with the help of this PR). The person could change the callback to the function that runs the query and then set the `$variations` to null, just before the API call. This way `get_variations` would execute the custom callback.


Filters plays a role majorly when we want to retain some part of the original variations; if not replacing the callback is more performant.

joemcgill replied:
Thanks for sharing the issue for a generic PHP helper for registering variations on a block, @gziolo. While I do think that being able to filter the list of registered variations on a block makes this easier, what is being requested in https://github.com/WordPress/gutenberg/issues/47170 is essentially just a utility function that would allow you to add a new variation object to the array stored in `WP_Block_Type::variations`, since currently the variations property essentially acts as a registry.

We could add a new `register_variation` method to this object that could be used to append a new variation to the list. The trick would be that we would need to ensure that any `variation_callback` declared when the block was registered gets run prior to adding any new variations, or temporarily store newly registered variations in a separate property so their presence doesn't keep the callback from running the first time the variations property is accessed nor cause the overhead from those callbacks to be called unnecessarily. Of the two options, I'd prefer the latter.

gziolo replied:
Introducing a new helper method is definitely a separate issue, but I appreciate your feedback on potential ways to implement it taking performance implications into consideration 👍

For what it's worth a similar helper for registering block styles prints instead JS code that moves the processing to the client and accounts fot custom block that weren’t registered on the server😀

spacedmonkey commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -540,4 +563,30 @@ public function get_attributes() {
+	 * @return array[]
+	 */
+	public function get_variations() {
+		if ( ! isset( $this->variations ) ) {
+			$this->variations = array();
+			if ( is_callable( $this->variation_callback ) ) {
+				$this->variations = call_user_func( $this->variation_callback );
+			}
+		}
```
```suggestion
		if ( $this->variation_callback && is_callable( $this->variation_callback ) ) {
			$this->variations = call_user_func( $this->variation_callback );
			$this->variation_callback = null;
		}
```

Another way of doing this would like this.

spacedmonkey replied:
Looking at other feedback, but there maybe a benefit of existing solution, as it keeps `variation_callback`.

kt-12 replied:
@spacedmonkey I had given this a thought but then I was thinking what if the developer needs the original callback name in the filter provided below?

kt-12 replied:
Like in  https://github.com/WordPress/gutenberg/pull/54801, here the problem is that variations are built a bit early. In the same way, there may be a case where the developer wants to run the original callback but at a later stage.

spacedmonkey replied:
Fair point. Works for me.

spacedmonkey commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -353,6 +367,10 @@ public function __get( $name ) {
 	 *              or false otherwise.
 	 */
 	public function __isset( $name ) {
+		if ( 'variations' === $name && isset( $this->variation_callback ) ) {
```
```suggestion
		if ( 'variations' === $name && ( isset( $this->variation_callback ) || isset( $this->variations ) ) ) {
```

I think this is needed, note it is a private property.

spacedmonkey commented:
@kt-12 @joemcgill We should add a backwards compablitiy shim in the gutenberg plugin. Something like this should work. 

```
function gutenberg_register_block_type_args_shim( $args ) {
    if( isset( $args['variation_callback'] ) && is_callable( $args['variation_callback'] ) ) {
           $args['variations'] = call_user_func( $args['variation_callback'] );
           unset( $args['variation_callback'] );
    } 
    return $args;
}
add_filter( 'register_block_type_args', 'gutenberg_register_block_type_args_shim' );
```

kt-12 committed:
isset logic

spacedmonkey reviewed (changes requested):
Let's revert the changes to the block files. These changes along with a backwards compatiblity shim, should be made in gutenberg project. 
 Can we also add a unit test for the `get_variations` method? 

Otherwise, this is good to go.

kt-12 commented:
> Let's revert the changes to the block files. These changes along with a backwards compatiblity shim, should be made in gutenberg project. Can we also add a unit test for the `get_variations` method?
> 
> Otherwise, this is good to go.

Yes, I am working on the unit test and Gutenberg changes. I'll add an update shortly.

felixarntz reviewed (commented):
@kt-12 This is a smart, clean, and backward compatible solution - excellent work!

I think once the block code is properly split out and implemented via Gutenberg, we can probably commit this.

kt-12 committed:
test cases for variation callback

kt-12 committed:
remove isset as variations will never be returned as null.

kt-12 committed:
isset always true for variations

kt-12 commented:
@joemcgill @spacedmonkey I had to replace `__isset` logic to always return true for `variations` to solve a BC issue - https://github.com/WordPress/wordpress-develop/actions/runs/7461317691/job/20301216384?pr=5718#step:14:340
If we think about it `variations` will never be returned as `null` due to the logic inside `get_variations`. Any call to `isset` inside `class WP_Block_Type` will never go through the magic function so it doesn't affect `isset` logic inside `get_variations`

To Note: With the current change, we introduce a BC issue (with a lower chance of re-surfacing); we can no longer set `variations` to `null`.

spacedmonkey commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -353,6 +367,10 @@ public function __get( $name ) {
 	 */
 	public function __isset( $name ) {
+		if ( 'variations' === $name ) {
+			return true;
```
```suggestion
			return ( null !== $this->variations || null !== $this->variation_callback );
```

Does this work?

kt-12 replied:
No, this is running into an error for some older logic. 
`null !== $this->variations` is the problem.
The issue is with the `default` value,  your suggested `isset` logic returns `false` as the default for `$variations` is currently `null`.  However, this contradicts to code in the trunk as it used to return `true` given the default value for `$variaitons` was `array()`.

To match the default value for B/C, we are returning `array()` if anyone tries to fetch variations directly. 
But, with this `isset` logic in place, we were getting different values for `$variations` than expected in some of the testcases.
For e.g this place here - https://github.com/10up/wordpress-develop/blob/9225964063a0609de92268fea28d0aed66d86542/src/wp-admin/includes/post.php#L2323
Here earlier we used to get `isset( $block_type->variations )` as `true` and now we were getting ` isset( $block_type->variations )` as `false`, even though in both cases `$block_type->variations` returned `array()` as value.

See this error - https://github.com/WordPress/wordpress-develop/actions/runs/7461317691/job/20301216384?pr=5718#step:14:340

This is the reason why I switched to returning `true` always, which is kind of correct according to the new logic. 
See the code snippet below. 
```php
// This does set variations to null
$block_type->variations = null;

// However if I fetch variations immediately I'll get a blank array. 
$value =  $block_type->variations;  // $block_type->get_variations();
print_r( $value ); // this will print `array()`

// so isset must return true as practically variations can never be null, atleast for external entity
```

This opens up a new issue, which is we can never set `$variations` to null and run external logic based on a null value.

Now the open question is - 
Do we need the default value as `array()` or do we need the ability to set `variations` to `null`. It looks like we can't have both privileges.

joemcgill replied:
Thanks for the explanation, @kt-12. This is definitely unintuitive, but does make sense based on the current expected default value for $block_type->variations being an empty array. I can't think of a use case where someone would want to manually set that property to a null value after it was instantiated, and that would likely break things anyway since the property is documented to be an array always, so I think what you're proposing is fine.

spacedmonkey replied:
Thanks for the feedback. Seems right to be.

spacedmonkey commented:
> Yes, I am working on the unit test and Gutenberg changes. I'll add an update shortly.

Is this still being worked on?

kt-12 commented:
> > Yes, I am working on the unit test and Gutenberg changes. I'll add an update shortly.
> 
> Is this still being worked on?

@spacedmonkey I did add a couple of test cases here in this PR and also updated GB PR https://github.com/WordPress/gutenberg/pull/56952

kt-12 committed:
revert back navigation-link to

kt-12 committed:
revert back to trunk

kt-12 committed:
Merge branch 'trunk' into enhancement/remove-variation-fontend

spacedmonkey reviewed (approved):
:shipit:

Mamaduka commented:
Excellent work, everyone!

joemcgill commented on tests/phpunit/tests/blocks/wpBlockType.php:
```diff
@@ -460,4 +460,47 @@ public function data_block_version() {
+		);
+
+		$this->assertSameSets( $this->mock_variation_callback(), $block_type->get_variations() );
+	}
```
We should probably add a few additional unit tests here to account for cases where the variations are already set, and when something is hooked into the variations callback to demonstrate the expected behavior that the callback will only be run on the first iteration and that manually set variations will override the existence of a variations callback if not already called, but those can be added in an additional commit if needed.

kt-12 replied:
I'll add it first thing in the morning tomorrow so we can have it in the final merge.

kt-12 replied:
@joemcgill I have added 7 more test cases to cover such cases.

joemcgill reviewed (approved):
This looks good, @kt-12! I left an inline comment about us adding some more unit tests for various use cases, but can commit this as is in the meantime.

mukeshpanchal27 commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -113,9 +113,18 @@ class WP_Block_Type {
+	 * Block variations callback.
+	 *
+	 * @since 6.5.0
+	 * @var callable
```
Initial state for the property is `null` why we don't added it here?

kt-12 replied:
Thanks for the catch. I have added `null` to the definition

mukeshpanchal27 commented on src/wp-includes/class-wp-block-type.php:
```diff
@@ -540,4 +563,30 @@ public function get_attributes() {
+		 * @param array         $variations Array of registered variations for a block type.
+		 * @param WP_Block_Type $block_type The full block type object.
+		 */
+		return apply_filters( 'get_block_type_variations', $this->variations, $this );
```
Good if you add unit test for new filter.

kt-12 replied:
@mukeshpanchal27  I have added 2 test cases to cover this.

kt-12 committed:
set to null

kt-12 committed:
test cases for multiple cases

kt-12 committed:
additional test cases

kt-12 committed:
change message

kt-12 committed:
cs fixes

mukeshpanchal27 commented on tests/phpunit/tests/blocks/wpBlockType.php:
```diff
@@ -460,4 +460,189 @@ public function data_block_version() {
+	 * @ticket 59969
+	 */
+	public function test_get_block_type_variations_filter_variations() {
+		// Filter will override the variations set during resistration.
```
```suggestion
		// Filter will override the variations set during registration.
```

kt-12 committed:
Update tests/phpunit/tests/blocks/wpBlockType.php

Fix typo

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

mukeshpanchal27 commented on tests/phpunit/tests/blocks/wpBlockType.php:
```diff
@@ -460,4 +460,189 @@ public function data_block_version() {
+
+		$this->assertSame( true, $callback_called, 'The callback should be called when the variations are accessed.' );
+		$this->assertSameSets( $obtained_variations, $expected_variations, 'The variations obtained from the callback should be filtered.' );
+	}
+	public function filter_test_variations( $variations, $block_type ) {
```
```suggestion
	}

	public function filter_test_variations( $variations, $block_type ) {
```

kt-12 committed:
Update tests/phpunit/tests/blocks/wpBlockType.php

New line

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

joemcgill commented on tests/phpunit/tests/blocks/wpBlockType.php:
```diff
@@ -460,4 +460,190 @@ public function data_block_version() {
+		);
+
+		$obtained_variations = $block_type->variations; // access the variations.
+		remove_filter( 'get_block_type_variations', array( $this, 'filter_test_variations' ), 10 );
```
Hooks get reset in `WP_UnitTestCase_Base::set_up` that runs before each test ([link](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/includes/abstract-testcase.php#L111-L113)), so no need to clean up filters manually in tests.

🔢 Applies here and throughout.

kt-12 replied:
Done.

joemcgill commented on tests/phpunit/tests/blocks/wpBlockType.php:
```diff
@@ -460,4 +460,190 @@ public function data_block_version() {
+		$this->assertSameSets( $obtained_variations, $expected_variations, 'The variations obtained from the callback should be filtered.' );
+	}
+
+	public function filter_test_variations( $variations, $block_type ) {
```
Nit: Can you move this function so it's defined prior to all of the to before all of the tests that make use of this filter? Also, adding a docblock here would be terrific.

kt-12 replied:
Done.

joemcgill reviewed (changes requested):
Thanks @kt-12. The additional tests look great! Just a couple of small nitpicks and I think this is about ready.

kt-12 committed:
Move filter function one up

kt-12 committed:
remove filters and expand doc block defnetion

joemcgill reviewed (approved):
The updates look great. Thanks, @kt-12!

joemcgill commented:
Merged in [57315](https://core.trac.wordpress.org/changeset/57315)

ellatrix commented:
🚨 PHP unit tests are currently failing in the GB repository. It started happening around the time this PR was merged.

```
Indirect modification of overloaded property WP_Block_Type::$variations has no effect
```

Example run: https://github.com/WordPress/gutenberg/actions/runs/7590355231/job/20676711107?pr=58031

kt-12 commented:
@ellatrix Thank you for flagging this. I have found the issue and have a fix for it. I'll push it once I test it out for other scenarios.

ellatrix commented:
Thanks!

kt-12 commented:
> @ellatrix Thank you for flagging this. I have found the issue and have a fix for it. I'll push it once I test it out for other scenarios.

I have added a new ticket for the bug with a detailed description - https://core.trac.wordpress.org/ticket/60309#ticket

I have added an initial fix https://github.com/WordPress/wordpress-develop/pull/5915. However, I need to check how my changes affect some operations in `__isset`
